### PR TITLE
[5.1] Dependency injection in model factories

### PR DESCRIPTION
**Pull request by dan-har**
It can be useful to inject some other classes to the factories, like some custom generators for data that is not included in `Faker`

example of use : 

```
$factory->define(App\User::class, function (App\CustomGenerator $customGenerator, $faker, $attributes) {

    return [
        'name' => $faker->name,
        'email' => $faker->email,
        'password' => str_random(10),
        'remember_token' => str_random(10),
        'custom_attribute' => $customGenerator->someValue()
    ];
});
```

Also, the function `Factory::raw()` was not compatible with the change in #9387, so it was fixed.

### CODE REVIEW

**Comment by Sladewill** [comment]
That's a breaking change for anyone using that currently.

**Comment by dan-har** [comment]
Can you explain why?

**Comment by Sladewill** [comment]
Because the first expected parameter is supposed to be faker, anyone currently using that it will break for.

**Comment by dan-har** [comment]
The good thing about the method dependency injection in the framework is that it take care about this, so you decide if you need more dependencies or not. So you can use it like that: 
```
$factory->define(App\User::class, function ($faker, $attributes) {
...
}
```

or even like that 

```
$factory->define(App\User::class, function (CustomGenerator $customGenerator, OtherGenerator $otherGenerator, $faker, $attributes) {
...
}
```

and it will still work as expected, and you will be able to use the `Faker`

Look at the [`call` method on the container](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Container/Container.php#L494) and on the [line responsible for the order of the parameters injection](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Container/Container.php#L535)

**Comment by Sladewill** [comment]
Yes but people are already using it with $faker as parameter 1 and attributes as parameter 2 making this a breaking change.

**Comment by dan-har** [comment]
attributes as the second parameter was added before, but it does not appear in the documentation ( in my opinion this parameter is not required..)

see #9387

**Comment by devinfd** [comment]
As the author of the PR https://github.com/laravel/framework/pull/9387 I guess I should chime in. Sending the attributes through is very helpful if, for example, you are defining relationships in the factory definition and you need to override the foreign key. ex:
```php
$factory->define(App\User::class, function ($faker, $attributes) {

    $accountId = isset($attributes['account_id']) ? $attributes['account_id']: factory(App\Account::class)->create()->id;

    return [
        'account_id' => $accountId,
        'email' => $faker->email,
        'password' => 'secret',
        'name' => $faker->name
    ];
});
```
Another example (using a real life scenario that I have) is to only set a value on the model if an attributes value meets a certain condition:
```php
$factory->define(App\Animal::class, function ($faker, $attributes) {
    return [
        'common_name' =>  $faker->word,
        'alpha_code' => $attributes['class'] === 'aves' ? $faker->lexar('????') : null,
        'is_mbta' =>$attributes['class'] === 'aves' ? $faker->randomElement([1,0]) : 0,
    ];
});
```
Basically the point is to keep the factory more informed about what needs to get created. I did submit a PR for documentation on this feature a while back but it was denied.

File: src/Illuminate/Database/Eloquent/Factory.php [hunk 1]
```diff
@@ -18,11 +26,13 @@ class Factory implements ArrayAccess
     /**
      * Create a new factory instance.
      *
+     * @param  \Illuminate\Contracts\Container\Container $container
      * @param  \Faker\Generator  $faker
      * @return void
      */
-    public function __construct(Faker $faker)
+    public function __construct(Container $container, Faker $faker)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Factory.php]: this is breaking

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Factory.php]: please put the order the other way arround.

File: src/Illuminate/Database/Eloquent/Factory.php [hunk 2]
```diff
@@ -18,11 +26,13 @@ class Factory implements ArrayAccess
     /**
      * Create a new factory instance.
      *
+     * @param  \Illuminate\Contracts\Container\Container $container
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Factory.php]: 2 spaced needed, not 1

File: src/Illuminate/Database/Eloquent/Factory.php [hunk 3]
```diff
@@ -36,15 +46,16 @@ public function __construct(Faker $faker)
...
      *
-     * @param  \Faker\Generator  $faker
-     * @param  string|null  $pathToFactories
+     * @param  \Illuminate\Contracts\Container\Container $container
+     * @param  \Faker\Generator $faker
+     * @param  string|null $pathToFactories
      * @return static
      */
-    public static function construct(Faker $faker, $pathToFactories = null)
+    public static function construct(Container $container, Faker $faker, $pathToFactories = null)
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Factory.php]: again, this is breaking

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Factory.php]: not sure how we can get around this

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Factory.php]: this probably need to go to 5.2 instead

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 4]
```diff
@@ -3,11 +3,19 @@
 namespace Illuminate\Database\Eloquent;
 
 use Faker\Generator as Faker;
+use Illuminate\Contracts\Container\Container;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: sort by length please

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 5]
```diff
@@ -45,16 +53,17 @@ class FactoryBuilder
...
      * Create an new builder instance.
      *
-     * @param  string  $class
-     * @param  string  $name
-     * @param  array  $definitions
-     * @param  \Faker\Generator  $faker
-     * @return void
+     * @param  string $class
+     * @param  string $name
+     * @param  \Illuminate\Contracts\Container\Container $container
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: cs

File: src/Illuminate/Database/Eloquent/FactoryBuilder.php [hunk 6]
```diff
@@ -45,16 +53,17 @@ class FactoryBuilder
...
+     * @param  array $definitions
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/FactoryBuilder.php]: cs

**Comment by dan-har** [comment]
@devinfd different optnions, but we are on the same page :+1:

**Comment by dan-har** [comment]
@GrahamCampbell The container parameter in the constructors can be changed so it wont be breaking, so the results will be that the container will be null as default and the factory closure will be called like this:
```
if($this->container){
    $definition = $this->container->call($this->definitions[$this->class][$this->name], [$this->faker, $attributes]);
} else {
    $definition = call_user_func($this->definitions[$this->class][$this->name], $this->faker, $attributes);
}
```

what do you think?

**Comment by GrahamCampbell** [comment]
Send to 5.2 please.
